core/corehttp/gateway_handler.go

```diff
@@ -273,116 +275,82 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
+	rootPath, err := path.ParsePath(r.URL.Path)
 	if err != nil {
-		webError(w, "Could not recursively add new node", err, http.StatusInternalServerError)
+		webError(w, "putHandler: ipfs path not valid", err, http.StatusBadRequest)
 		return
 	}
 
-	i.addUserHeaders(w) // ok, _now_ write user's headers.
-	w.Header().Set("IPFS-Hash", key.String())
-	http.Redirect(w, r, ipfsPathPrefix+key.String()+"/", http.StatusCreated)
-}
-
-func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
-	// TODO(cryptix): either ask mildred about the flow of this or rewrite it
-	webErrorWithCode(w, "Sorry, PUT is bugged right now, closing request", errors.New("handler disabled"), http.StatusInternalServerError)
-	return
-	urlPath := r.URL.Path
-	pathext := urlPath[5:]
-	var err error
-	if urlPath == ipfsPathPrefix+"QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn/" {
-		i.putEmptyDirHandler(w, r)
+	rsegs := rootPath.Segments()
+	if rsegs[0] == ipnsPathPrefix {
+		webError(w, "putHandler: updating named entries not supported", errors.New("WritableGateway: ipns put not supported"), http.StatusBadRequest)
 		return
 	}
 
 	var newnode *dag.Node
-	if pathext[len(pathext)-1] == '/' {
+	if rsegs[len(rsegs)-1] == "QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn" {
 		newnode = uio.NewEmptyDirectory()
 	} else {
-		newnode, err = i.newDagFromReader(r.Body)
+		putNode, err := i.newDagFromReader(r.Body)
 		if err != nil {
-			webError(w, "Could not create DAG from request", err, http.StatusInternalServerError)
+			webError(w, "putHandler: Could not create DAG from request", err, http.StatusInternalServerError)
 			return
 		}
+		newnode = putNode
 	}
 
-	ctx, cancel := context.WithCancel(i.node.Context())
-	defer cancel()
-
-	ipfsNode, err := core.Resolve(ctx, i.node, path.Path(urlPath))
-	if err != nil {
-		// FIXME HTTP error code
-		webError(w, "Could not resolve name", err, http.StatusInternalServerError)
-		return
-	}
-
-	k, err := ipfsNode.Key()
-	if err != nil {
-		webError(w, "Could not get key from resolved node", err, http.StatusInternalServerError)
-		return
-	}
-
-	h, components, err := path.SplitAbsPath(path.FromKey(k))
-	if err != nil {
-		webError(w, "Could not split path", err, http.StatusInternalServerError)
-		return
+	var newPath string
+	if len(rsegs) > 1 {
+		newPath = strings.Join(rsegs[2:], "/")
 	}
 
-	if len(components) < 1 {
-		err = fmt.Errorf("Cannot override existing object")
-		webError(w, "http gateway", err, http.StatusBadRequest)
-		return
-	}
-
-	tctx, cancel := context.WithTimeout(ctx, time.Minute)
-	defer cancel()
-	// TODO(cryptix): could this be core.Resolve() too?
-	rootnd, err := i.node.Resolver.DAG.Get(tctx, key.Key(h))
-	if err != nil {
-		webError(w, "Could not resolve root object", err, http.StatusBadRequest)
-		return
-	}
+	var newkey key.Key
+	rnode, err := core.Resolve(i.node.Context(), i.node, rootPath)
+	switch ev := err.(type) {
```

Comment:
This is a switch from err casting.
This is cleaner, and should be used across the code.

In current master, where cast is used:
```shell
$ grep -rE 'err\.\(' . | grep -v Godeps
./routing/dht/ext_test.go:		if merr, ok := err.(u.MultiErr); ok && len(merr) > 0 {
./routing/dht/ext_test.go:		if merr, ok := err.(u.MultiErr); ok && len(merr) > 0 {
./routing/dht/ext_test.go:		if merr, ok := err.(u.MultiErr); ok && len(merr) > 0 {
./core/corehttp/gateway_handler.go:	if _, ok := err.(path.ErrNoLink); ok {
./core/corehttp/gateway_handler.go:	if _, ok := err.(path.ErrNoLink); ok {
./core/commands/add.go:		if _, ok := err.(*hiddenFileError); ok {
./p2p/net/conn/dial_test.go:			if ne, ok := err.(net.Error); ok && ne.Temporary() {
./p2p/net/conn/dial.go:	if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
./p2p/net/conn/dial.go:	errno, ok := err.(syscall.Errno)
./cmd/ipfs/main.go:	switch e := err.(type) {
```